### PR TITLE
Feature/SOF-7360 feat: add simple defect builder

### PR DESCRIPTION
**Pull request by VsevolodX**

### CODE REVIEW

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
+from typing import List
+
+from pydantic import BaseModel
+from pymatgen.analysis.defects.core import Substitution, Vacancy, Interstitial
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:4]: Let's prepend Pymatgen here too

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from pymatgen.core import PeriodicSite
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:5]: And here

File: src/py/mat3ra/made/tools/build/defect/builders.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            defect = Substitution(structure=pymatgen_structure, site=pymatgen_periodic_site)
+        elif configuration.defect_type == PointDefectTypeEnum.INTERSTITIAL:
+            defect = Interstitial(structure=pymatgen_structure, site=pymatgen_periodic_site)
+        else:
+            raise ValueError(f"Unknown defect type: {configuration.defect_type}")
+
+        return [defect.centered_defect_structure if self.build_parameters.center_defect else defect.defect_structure]
+
+    @staticmethod
+    def _convert_generated_item(item: _GeneratedItemType):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/builders.py:51]: Mixin??

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    # TODO: fix arbitrary_types_allowed error and set Material class type
+    material: Any = None
+    # defect_type type can be an Enum for a specific defect class (for point defect, 2d defect, etc.)
+    defect_type: Union[PointDefectTypeEnum, None] = None
+
+
+class PointDefectConfiguration(BaseDefectConfiguration):
+    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.VACANCY
+    # TODO: should come from Enum of elements
+    specie: Optional[str] = None
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:18]: Or from https://github.com/Exabyte-io/periodic-table

Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/defect/configuration.py:18]: Ok. We didn't depend on that before

File: src/py/mat3ra/made/tools/build/defect/__init__.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from typing import Union, Optional
+
+from mat3ra.made.material import Material
+from .builders import VacancyBuilder, SubstitutionBuilder, InterstitialBuilder, PointDefectBuilderParameters
+from .configuration import VacancyConfiguration, SubstitutionConfiguration, InterstitialConfiguration
+from .enums import PointDefectTypeEnum
+
+
+def DEFECT_BUILDER_MAP(builder_parameters):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/__init__.py:9]: This is a factory

File: src/py/mat3ra/made/tools/build/defect/configuration.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+class VacancyConfiguration(BasePointDefectConfiguration):
+    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.VACANCY
+
+
+class SubstitutionConfiguration(BasePointDefectConfiguration):
+    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.SUBSTITUTION
+    element: Optional[str] = "Si"
+
+
+class InterstitialConfiguration(BasePointDefectConfiguration):
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/configuration.py:45]:
```
class BaseDefectConfiguration(BaseModel):
    material: Any = None
    defect_type: Union[PointDefectTypeEnum, None] = None


class BasePointDefectConfiguration(BaseDefectConfiguration):
    # In what units?
    position: Optional[List[float]] = [0, 0, 0]
    site_id: Optional[int] = 0

    def _get_site_by_coordinates(self):
        # Get the closest site to the self.position
        pass


class VacancyPointDefectConfiguration(BasePointDefectConfiguration):
    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.VACANCY


class SubstitutionPointDefectConfiguration(BasePointDefectConfiguration):
    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.SUBSTITUTION
    element: Optional[str] = 'Si'


class InterstitialPointDefectConfiguration(BasePointDefectConfiguration):
    defect_type: PointDefectTypeEnum = PointDefectTypeEnum.INTERSTITIAL
    element: Optional[str] = 'Si'
```

**Review by timurbazhirov — APPROVED** [review]

File: src/py/mat3ra/made/material.py [hunk 7]
```diff
@@ -3,6 +3,7 @@
 from mat3ra.code.constants import AtomicCoordinateUnits, Units
 from mat3ra.code.entity import HasDescriptionHasMetadataNamedDefaultableInMemoryEntity
 from mat3ra.esse.models.material import MaterialSchema
+from mat3ra.made.utils import map_array_with_id_value_to_array
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/material.py:6]: I thought we are using https://github.com/Exabyte-io/made/blob/main/src/py/mat3ra/made/utils.py

File: src/py/mat3ra/made/tools/build/defect/__init__.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+from mat3ra.made.material import Material
+
+from .builders import PointDefectBuilderParameters
+from .configuration import PointDefectConfiguration
+from .enums import PointDefectTypeEnum
+
+
+class DefectBuilderFactory(BaseFactory):
+    __class_registry__ = {
+        PointDefectTypeEnum.VACANCY: "mat3ra.made.tools.build.defect.builders.VacancyPointDefectBuilder",
```
Comment (timurbazhirov) [inline src/py/mat3ra/made/tools/build/defect/__init__.py:13]: Does this understand local references `.builders.Vacancy...`?

File: tests/py/unit/test_tools_build_defect.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    assert len(defect.basis["elements"]) == 1
+
+
+def test_create_substitution():
+    # Substitution of Ge in place of Si at default site_id=0
+    configuration = PointDefectConfiguration(crystal=clean_material, defect_type="substitution", chemical_element="Ge")
+    defect = create_defect(configuration)
+
+    assert defect.basis["elements"] == [{"id": 0, "value": "Ge"}, {"id": 1, "value": "Si0+"}]
```
Comment (timurbazhirov) [inline tests/py/unit/test_tools_build_defect.py:20]: What is this `Si0+`?? The elements should have actual element names only without anything else

File: src/py/mat3ra/made/material.py (hunk 7 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/material.py:6]: mat3ra.made.utils?

File: src/py/mat3ra/made/tools/build/defect/__init__.py (hunk 8 above)
Comment (VsevolodX) [reply src/py/mat3ra/made/tools/build/defect/__init__.py:13]: Looks like no.
```
tests/py/unit/test_tools_build_defect.py::test_create_vacancy - 
TypeError: the 'package' argument is required to perform a relative import for '.builders'
```

**Review by timurbazhirov — APPROVED** [review]

**Pull request merged**